hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ContainerStreamingSource.java

```diff
@@ -0,0 +1,110 @@
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.container.replication;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.hadoop.hdds.protocol.datanode.proto.ContainerProtos.ContainerDataProto.State;
+import org.apache.hadoop.hdds.scm.container.common.helpers.StorageContainerException;
+import org.apache.hadoop.ozone.container.common.impl.ContainerSet;
+import org.apache.hadoop.ozone.container.keyvalue.KeyValueContainer;
+import org.apache.hadoop.ozone.container.stream.StreamingException;
+import org.apache.hadoop.ozone.container.stream.StreamingSource;
+
+/**
+ * Streaming source for closed-container replication.
+ */
+public class ContainerStreamingSource implements StreamingSource {
+
+  private ContainerSet containerSet;
+
+  public ContainerStreamingSource(ContainerSet containerSet) {
+    this.containerSet = containerSet;
+  }
+
+  @Override
+  public Map<String, Path> getFilesToStream(String id) {
+
+    Map<String, Path> filesToStream = new HashMap<>();
+
+    final long containerId = Long.parseLong(id);
+    final KeyValueContainer container =
+        (KeyValueContainer) containerSet.getContainer(containerId);
+
+
+    if (container.getContainerState() != State.CLOSED
+        && container.getContainerState() != State.QUASI_CLOSED) {
+      throw new StreamingException(
+          "Only (quasi)closed containers can be exported, but ContainerId=: "
+              + id + " is in state " + container.getContainerState());
+    }
+
+    try {
+      container.release();
+    } catch (StorageContainerException e) {
+      throw new StreamingException("Container couldn't be released: " + id, e);
+    }
+
+
+    try {
+
+      final File dbPath =
+          container.getContainerData().getContainerDBFile();
+
+      List<Path> dbFiles;
+      try (Stream<Path> files = Files.list(dbPath.toPath())) {
+        dbFiles = files.collect(Collectors.toList());
+      }
+
+      for (Path dbFile : dbFiles) {
+        if (dbFile.getFileName() != null) {
+          filesToStream.put("DB/" + dbFile.getFileName(), dbFile);
+        }
+      }
+
+      filesToStream.put("container.yaml",
+          container.getContainerData().getContainerFile().toPath());
+
+      final Path dataPath =
+          Paths.get(container.getContainerData().getChunksPath());
+
+      try (Stream<Path> walk = Files.walk(dataPath)) {
+
+        walk.filter(Files::isRegularFile)
+            .forEach(path -> {
+              filesToStream
+                  .put("DATA/" + dataPath.relativize(path), path);
+            });
+      }
+      
+    } catch (IOException e) {
+      throw new StreamingException("Couldn't stream container " + containerId,
+          e);
+    }
+    return filesToStream;
```

Comment:
There is a lock protection during the whole container transfer phase currenlty，to prevent concurrent modification of the client， say delete some blocks.  

Now this lock is removed.  I'm more sure If the file in filesToStream gets deleted before transmitted to server, what will be the consequence?  How to guarantee that the metadata and chunk data is consistent.